doc/man/man1/patt.1
\"    This documentation is copyright 1994 David Huron.
.TH patt 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +10
.ti -10
\fBpatt\fR  --  locate and output user-defined patterns in a Humdrum input
.in -10
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
.in +6
.ti -6
\fBpatt\fR  [\fB-m\fR]  \fB-f \fItemplatefile\fR  [\fB-s \fIregexp\fR]  [\fB-t \fIoutput_tag\fR]  [\fIinputfile\fR ...]
.in -6
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "patt"
command is used to locate occurrences of
a user-defined pattern in some Humdrum input.
The patterns sought may span innumerable data records.
Occurrences of the pattern may be identified by line number,
echoed intact in the output stream, or tagged by a user-defined marker
in a \f(CR**patt\fR output spine.
.sp 1
.sp 1
The pattern sought must be defined as a separate \(odtemplate\(cd file.
The template file is identified using the
.B "-f"
command option.
.sp 1
.sp 1
Pattern templates consist of one or more records.
Each record specifies a regular expression pattern.
The input is scanned from beginning to end, in order to
find passages that match the defined template.
In order for a match to take place, successive records in
the input stream must match the regular expressions given in each of
the corresponding records of the template.
For example, if the template is 9 lines in length,
then the input stream must contain 9 successive lines of matching
data in order for a pattern match to be successful.
.sp 1
.sp 1
The
.B "patt"
command implements a full \s-1UNIX\s+1 regular expression syntax.
Each line in the template file represents an independent regular
expression.
For example, the template:
.sp 1
.sp 1
.na
.in +2
.nf
1
2
3
.fi
.in -2
.SA
.sp 1
.sp 1
will match inputs such as the following:
.sp 1
.sp 1
.in +2
1
.br
112
.br
43.9
.in -2
.sp 1
.sp 1
or even
.in +2
0x.=%&1*
.br
Figure 32
.br
abc(...32...)
.in -2
A more circumspect regular expression template might look like this:
.sp 1
.sp 1
.na
.in +2
.nf
^1$
^2$
^3$
.fi
.in -2
.SA
.sp 1
.sp 1
(The caret (^) and dollar sign ($) are regular expression
.I "anchors"
that indicate the beginning of the record and end of the record
respectively.)
Refer to
.I "regexp (6)"
for further information regarding regular expressions.
Note that the related
.B "pattern"
command implements a more powerful regular expression syntax
for defining multi-record patterns.
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "patt"
command supports the following options:
.sp 1
.TS
l l.
\fB-c\fR	makes pattern-matching sensitive to comments
\fB-e\fR	echoes matched patterns in the output
\fB-f \fItemplatefile\fR	use pattern specified in \fItemplatefile\fR
\fB-h\fR	displays a help screen summarizing the command
	  syntax
\fB-m\fR	invokes collapsed multiple-record matching mode
\fB-s \fIregexp\fR	skip (ignore) data records containing the
	  defined regular expression
\fB-t \fIoutput_tag\fR	generate \f(CR**patt\fR output spine; tag each
	  occurrence of the pattern with the string
	  \fIoutput_tag\fR
.TE
.sp 1
Options are specified in the command line.
.sp 1
.sp 1
By default, the
.B "patt"
command is insensitive to the presence or absence of Humdrum comments.
Pattern searches may be made sensitive to occurrences of comments by
specifying the
.B "-c"
option.
.sp 1
.sp 1
In the default operation,
.B "patt"
outputs a Humdrum global comment for each pattern matched in the input.
Each comment identifies the line number in the input where the
found pattern begins.
.sp 1
.sp 1
With the
.B "-e"
option, each instance of the found pattern is echoed in the output.
Each output pattern is preceded by the appropriate exclusive interpretation(s)
and followed by appropriate spine-path terminator(s).
.sp 1
.sp 1
Certain types of data records may be ignored in the pattern-search
by invoking the
.B "-s"
(skip) option.
This option must be accompanied by a user-define regular expression.
All input data records matching the regular expression are ignored.
This option is useful, for example, in skipping null data tokens,
barlines, marked embellishment tones, or other types of data.
.sp 1
.sp 1
The
.B "-m"
option invokes a multiple record matching mode.
In this mode,
.B "patt"
attempts to match as many successive regular expressions in the template file
as possible for a given input record, before continuing with the next
input and template records.
In this way, a template file consisting of several records,
may possibly match a single input record.
(See \s-1EXAMPLES\s+1 below.)
.sp 1
.sp 1
The
.B "-t"
option causes
.B "patt"
to output a single spine of \f(CR**patt\fR output.
The user specifies an \fIoutput tag\fR (character string) on the command line.
Each instance of the found pattern causes the tag to be
output in the \f(CR**patt\fR spine at the position corresponding to
the onset of each found pattern.
(See \s-1EXAMPLES\s+1 below.)
Note that the
.B "-t"
and
.B "-e"
options are mutually exclusive.
.sp 1
.sp 1
Whatever options are invoked,
.B "patt"
always produces output that conforms to the Humdrum syntax.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
The following examples illustrate the operation of the
.B "patt"
command.
Consider the following target file and pattern template file:
.sp 1
.sp 1
.B "template file:"
.in +2
.sp 1
.TS
l.
1
2
3
.TE
.sp 1
.in -2
.sp 1
.sp 1
.B "target file:"
.in +2
.sp 1
.TS
l.
**foo
1
2
\(eq1
1 3
2 1
3 2
3
\(eq2
1
2 3
4
2 3 1
*-
.TE
.sp 1
.in -2
A simple search for the pattern template might use the command:
.sp 1
.sp 1
.in +2
patt -f template target
.in -2
.sp 1
.sp 1
Pattern matches are announced by outputting Humdrum global comments.
Given the above command, the following output would result:
.in +2
.sp 1
.TS
l.
!! Pattern found at line 5 of file input
!! Pattern found at line 6 of file input
.TE
.sp 1
.in -2
In the first instance, notice that
.B "patt"
is able to identify overlapping patterns.
Two instances of the 1-2-3 pattern are identified
beginning on the fifth and sixth lines respectively.
.sp 1
.sp 1
Note however, that the first instance of the pattern (beginning at line 2)
was not identified due to the interruption of the common system barline in
the fourth line.
The barlines can be ignored by invoking the
.B "-s"
option, followed by a regular expression that uniquely identifies the
records to be skipped -- in this case the equals sign.
The command:
.sp 1
.sp 1
.in +2
patt -s = -f template target
.in -2
.sp 1
.sp 1
would produce the following output:
.in +2
.sp 1
.TS
l.
!! Pattern found at line 2 of file input
!! Pattern found at line 5 of file input
!! Pattern found at line 6 of file input
.TE
.sp 1
.in -2
Actual instances of the pattern can be output by invoking the
.B "-e"
(echo) option:
.sp 1
.sp 1
.in +2
patt -e -s = -f template target
.in -2
.sp 1
.sp 1
The following output would result:
.in +2
.sp 1
.TS
l.
!! Pattern found at line 2 of file input
**foo
1
2
\(eq1
1 3
*-
!! Pattern found at line 5 of file input
**foo
1 3
2 1
3 2
*-
!! Pattern found at line 6 of file input
**foo
2 1
3 2
3
*-
.TE
.sp 1
.in -2
Notice that each pattern found is output as a self-contained Humdrum
output with initial exclusive interpretations and concluding
spine-path terminators.
If a single continuous output is desired, the
.B "rid -u"
command may be used to eliminate the duplicate interpretations.
.sp 1
.sp 1
Instead of outputting the individual patterns, the
.B "-t"
option may be used to output a spine that marks each instance
of the found pattern.
In the following command, the beginning of each occurrence of the pattern
is labelled in the \f(CR**patt\fR spine by the tag \(odone-two-three.\(cd
.sp 1
.sp 1
.in +2
patt -t one-two-three -s = -f template target
.in -2
.sp 1
.sp 1
The follow output would result:
.in +2
.sp 1
.TS
l l.
**foo	**patt
1	one-two-three
2	.
\(eq1	.
1 3	one-two-three
2 1	one-two-three
3 2	.
3	.
\(eq2	.
1	.
2 3	.
4	.
2 3 1	.
*-	*-
.TE
.sp 1
.in -2
For some tasks (such as the identification of tone-rows in 12-tone music),
nominally \(odsuccessive\(cd elements of the pattern may be collapsed within
a single sonority or record.
The
.B "-m"
option invokes a
.I "multiple record matching"
mode.
By way of example, the following command:
.sp 1
.sp 1
.in +2
patt -m -t theme123 -s = -f template target
.in -2
.sp 1
.sp 1
will produce the following output:
.in +2
.sp 1
.TS
l l.
**foo	**patt
1	theme123
2	.
\(eq1	.
1 3	theme123
2 1	theme123
3 2	.
3	.
\(eq2	.
1	theme123
2 3	.
4	.
2 3 1	theme123
*-	*-
.TE
.sp 1
.in -2
.sp 1
.sp 1
Note that in the above examples, the extensive capabilities
for defining complex regular expressions have not been used.
Refer to \fBregexp (6)\fR for further pertinent information.
.in -2
.sp 1
.sp 1
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBgrep\fR (\s-1UNIX\s+1), \fBegrep\fR (\s-1UNIX\s+1),
\fBpattern\fR (1), \fBregexp\fR (1),
\fBsimil\fR (1)
.in -2
.sp 1
.sp 1
.SH "WARNINGS"
.in +2
If a comment is present in the template pattern,
failing to specify the
.B "-c"
option will make pattern matching a logically impossibility.
.in -2
.sp 1
.sp 1
.SH "NOTE"
.in +2
The
.B "patt"
command differs from the related
.B "pattern"
command in the following ways:
(1)
.B "patt"
always produces output conforming to the Humdrum syntax whereas
.B "pattern"
never produces Humdrum output.
(2)
.B "pattern"
allows multi-record `wild cards' in the template file,
and so permits the creation of more sophisticated regular expressions.
(3) The
.B "pattern"
command does not directly provide an \(odecho\(cd option.
.in -2


toolkit/awk/patt.awk
#####################################################################
##                          PATT.AWK
# Programmed by:  David Huron     Date: Dec., 1989
#
# Modifications:
#   Date:      Programmer:    Description:
# 1993 Aug.5   David Huron    Added spine-path indicators
# 1993 Aug.19  David Huron    Added **patt output
# May 17/94	Tim Racinsky	Solidified options -- c, e, s, t, f
# June 7/94	Tim Racinsky	Modified to work with getopts
# June 20/94	Tim Racinsky	Added -m option (exhaustive search)
#
# The purpose of this program is to identify all occurrences of
# a user-defined pattern in the input stream.
#
# Main Functions used:
#	parse_command()		store_indicators()			store_new_interps()
#	process_indicators()	ins_array_pos()			del_array_pos()
#	exchange_spines()		reset_new()				reset_space()
#	process_comments()		process_interpretations()	process_pattern()
#	process_tag()			process_echo()				process_no_option()
#
#	VARIABLES:
#
#  -stderr:  variable holding destination of error messages
#  -first_file:  flag to indicate if this is the first file to be processed
#  -no_interpretation_yet:  flag to indicate if first interp. record processed
#  -current_no_of_spines:  holds current number of active spines
#  -null_interp:  Regular expression to match null interpretations
#  -spine_path_record:  Regular expression to match spine-path records
#  -spine_terminate:  Regular expression to match all spines terminating
#  -options:  Holds options specified by the user
#  -skip_reg:  Holds skip regular expression for -s option (if specified)
#  -output_tag:  Holds output tag for -t option (if specified)
#  -pattern_length:  Holds length of pattern to match
#  -token_length:  Holds current number of records read into token array
#  -buffer_length:  Holds current number of records read into buffer array
#  -pattern_start:  Pointer into buffer array showing where next possible 
#                   pattern may start
#  -interp_head:  Pointer showing where to insert next exclusive interpretation
#                 record into past_interp array (see past_interp definition)
#  -interp_tail:  Same as above except showing where to delete next
#  -buffer_start:  Pointer into buffer array after last printed entry
#
#	ARRAYS:
#
#  -path_indicator:  Holds spine-path indicator symbols for current spine-path
#                    record
#  -current_interp:  Holds exclusive interpretations for currently active spines
#  -past_interp:  Holds exclusive interpretations that were active at the time
#                 for the last set of records that may be part of a pattern.
#                 There are pattern_length + 1 elements and it is implemented as
#                 a limited queue -- thus, the header and tail pointers move
#                 mod (pattern_length + 1).  This array is necessary so that if
#                 a pattern is found, and the -e options was specified, the
#                 exclusive interpretations that were active when the pattern
#                 started can be printed.
#  -pattern:  Holds the pattern to match (read in from template file)
#  -token:  Holds the latest set of records to test against pattern array
#  -buffer:  Holds all records read in so that they may be output again.
#
BEGIN {
	USAGE="\nUSAGE: patt -h                   (Help Screen)\n       patt [-c"\
		 "em] [-t output_tag] [-s skip_reg] -f template [file ...]"
	TRUE = 1; FALSE = 0
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	#
	# Set regular expressions to be used in program
	#
	null_interp = "^\\*(\t\\*)*$"
	spine_path_record = "^\\*(\\+|v|\\^|x|-)?(\t\\*(\\+|v|\\^|x|-)?)*$"
	spine_terminate = "^\\*-(\t\\*-)*$"
	#
	# Parse command line options
	#
	no_interpretation_yet = TRUE
	current_no_of_spines = 0
	token_length = buffer_length = pattern_start = 0
	interp_head = interp_tail = 0
	buffer_start = 1
	first_file = TRUE
	options = ""
	skip_reg = ""
	output_tag = ""
	pattern_length = 0
	parse_command()
}
{
#
# If this is the first record of a file, reset variables and arrays
#
if (FNR == 1) reset_new()
#
# If the buffer has more than 5000 entries, clean it up
#
if (buffer_length > 5000) reset_space()
#
# Process the records according to their record type
#
if ($0 ~ /^!/ && options !~ /c/) process_comments()
else if ($0 ~ /^\*/) process_interpretations()
else process_pattern()
}
END {
	#
	# If -m option specified, there may be some more patterns to find in
	# the last few lines of data
	#
	if (options ~ /m/)
		{
		while (buffer_start <= buffer_length)
			{
			if (options ~ /t/) process_tag()
			else if (options ~ /e/) process_echo()
			else process_no_option()
			}
		}
	#
	# Be sure to flush the buffer if -t option specified
	#
	else if (options ~ /t/)
		{
		for (i = buffer_start; i <= buffer_length; i += 1) print buffer[i]
		}
	}

############################################################################
#				FUNCTIONS USED IN PATT.AWK
############################################################################

############################################################################
##				Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	#
	# Otherwise, process any necessary variables
	#
	if (ARGV[2] != "null") options = ARGV[2]
	#
	# Store the skip option and value if specified
	#
	if (ARGV[3] != "null")
		{
		skip_reg = ARGV[3]
		options = options "s"
		}
	#
	# Store the tag option and value if specified
	#
	if (ARGV[4] != "null")
		{
		output_tag = ARGV[4]
		options = options "t"
		}
	if (options ~ /t/ && options ~ /e/)
		{
		print "patt: ERROR: Cannot specify both -t and -e." > stderr
		print USAGE > stderr
		exit
		}
	#
	# Read in the "template" file.
	#
	while (getline < ARGV[5] > 0) pattern[++pattern_length] = $0
	ARGV[1] = ARGV[2] = ARGV[3] = ARGV[4] = ARGV[5] = ""
	}

############################################################################
##				Function reset_new
#	This function resets the variables when processing a new file
#
function reset_new(   i)
	{
	#
	# If this isn't the first file, flush the buffer
	#
	if (!first_file)
		{
		if (options ~ /t/)
			{
			for (i = buffer_start; i <= buffer_length; i += 1)
				print buffer[i]
			}
		}
	else first_file = FALSE
	#
	# Reset all variables and delete arrays
	#
	token_length = buffer_length = pattern_start = 0
	interp_head = interp_tail = 0
	buffer_start = 1
	for (i in buffer) delete buffer[i]
	for (i in token) delete token[i]
	no_interpretation_yet = TRUE
	current_no_of_spines = 0
	if (FILENAME == "-") file_name = "STNDIN"
	else file_name = FILENAME
	}

############################################################################
##				Function reset_space
#	This function resets the variables when the buffer array becomes too large
#
function reset_space(    i,counter,max_length)
	{
	counter = 0
	max_length = buffer_length
	#
	# Move the unprinted buffer information back to the start of the array
	#
	for (i = buffer_start; i <= buffer_length; i += 1)
		{
		buffer[++counter] = buffer[i]
		}
	#
	# Reset the variables and delete unused array elements
	#
	pattern_start = 1 + pattern_start - buffer_start
	buffer_start = 1
	buffer_length = counter
	for (i = counter + 1; i <= max_length; i += 1) delete buffer[i]
	}

###########################################################################
##				Function process_comments
#	This function processes comments when the -c option is not specified. If
# the -t option was specified and this is a local comment, add a null local
# comment to the end of the record.  Always store it in the buffer.
#
function process_comments()
	{
	if (options ~ /t/ && $0 ~ /^![^!]/) $0 = $0 "\t!"
	buffer[++buffer_length] = $0
	}

############################################################################
##				Function process_interpretations
#	This function processes interpretation records
#
function process_interpretations()
	{
	#
	# If this is the first exclusive interpretation, store the interps., set
	# the variables and add the **patt spine if -t selected
	#
	if (no_interpretation_yet)
		{
		current_no_of_spines = NF
		no_interpretation_yet = FALSE
		store_new_interps()
		if (options ~ /t/) $0 = $0 "\t**patt"
		}
	else
		{
		#
		# If its a spine-path record, process the spine-path indicators
		#
		if ($0 ~ spine_path_record && $0 !~ null_interp)
			{
			store_indicators()
			process_indicators()
			if (current_no_of_spines == 0) no_interpretation_yet = TRUE
			}
		else store_new_interps()
		#
		# If the -t option specified, add the extra null interpretation or a
		# spine-path terminator if all other spines terminated
		#
		if (options ~ /t/)
			{
			if ($0 ~ spine_terminate) $0 = $0 "\t*-"
			else $0 = $0 "\t*"
			}
		}
	#
	# Store the record in the buffer
	#
	buffer[++buffer_length] = $0
	}

###########################################################################
##				Function process_pattern
#	This function processes the records which may make up a pattern
#
function process_pattern(   i,interp_line)
	{
	interp_line = ""
	#
	# If -t option, append a null token or null local comment
	#
	if (options ~ /t/)
		{
		if ($0 ~ /^![^!]/) $0 = $0 "\t!"
		else if ($0 !~ /^!!/) $0 = $0 "\t."
		}
	#
	# If -e option, store the exclusive interpretations for the current record
	# in past_interp.  If the array is full (number of elements = 
	# pattern_length, then delete the oldest element (at interp_tail)
	#
	else if (options ~ /e/)
		{
		for (i = 1; i <= current_no_of_spines; i += 1)
			{
			if (i == current_no_of_spines)
				interp_line = interp_line current_interp[i]
			else
				interp_line = interp_line current_interp[i] "\t"
			}
		if (((interp_head + 1) % (pattern_length + 1)) == interp_tail)
			interp_tail = (interp_tail + 1) % (pattern_length + 1)
		past_interp[interp_head] = interp_line
		interp_head = (interp_head + 1) % (pattern_length + 1)
		}
	#
	# Store the record and process if this record should not be skipped
	#
	buffer[++buffer_length] = $0
	if (options !~ /s/ || (options ~ /s/ && $0 !~ skip_reg))
		{
		#
		# Add the record to token array and set pattern_start pointer if
		# this is the first record of the first possible pattern
		#
		token[++token_length] = $0
		if (token_length == 1) pattern_start = buffer_length
		#
		# If there are enough records in token array, check for match
		#
		if (token_length == pattern_length)
			{
			if (options ~ /t/) process_tag()
			else if (options ~ /e/) process_echo()
			else process_no_option()
			}
		}
	}

#########################################################################
##				Function process_tag
#	This function will process the patterns as specified by -t option
#
function process_tag(    i,j,one_match,match_found)
	{
	#
	# Loop through the arrays to look for a pattern
	#
	match_found = TRUE
	#
	# If -m option was specified, then match as many patterns on each line
	# as is possible
	#
	if (options ~ /m/)
		{
		one_match = FALSE
		j = 1
		i = 1
		while (match_found && i <= pattern_length)
			{
			if (token[j] !~ pattern[i] && !one_match) match_found = FALSE
			else if (token[j] !~ pattern[i] && one_match)
				{
				j += 1
				one_match = FALSE
				}
			else
				{
				i += 1
				one_match = TRUE
				}
			}
		}
	else
		{
		for (i = 1; i <= pattern_length; i += 1)
			{
			if (token[i] !~ pattern[i]) { match_found = FALSE ; break }
			}
		}
	#
	# If they match, append the output tag to the end of the line
	#
	if (match_found)
		{
		if (buffer[pattern_start] ~ /^!!/)
			buffer[pattern_start] = buffer[pattern_start] " " output_tag
		else if (buffer[pattern_start] ~ /^!/)
			buffer[pattern_start] = buffer[pattern_start] output_tag
		else
			buffer[pattern_start] = substr(buffer[pattern_start],1,\
							length(buffer[pattern_start])-1) output_tag
		}
	#
	# Regardless, print any records up to the start of the current pattern
	# start record
	#
	for (i = buffer_start; i <= pattern_start; i += 1) print buffer[i]
	#
	# Determine where the next possible pattern could start and then reset
	# the token array
	#
	i = pattern_start + 1
	while ((buffer[i] ~ /^[*!]/) || (options ~ /s/ && buffer[i] ~ skip_reg))
		{
		if (options ~ /c/ && buffer[i] ~ /^!/) break
		print buffer[i++]
		}
	buffer_start = pattern_start = i
	for (i = 1; i <= token_length; i += 1) token[i] = token[i+1]
	token_length -= 1
	}

#########################################################################
##				Function process_echo
#	This function will process the patterns as specified by -e option
#
function process_echo(   i,j,one_match,match_found,terminate,spine_count\
								,data_count,temp_array,end_of_pattern)
	{
	#
	# Loop through the arrays to look for a pattern
	#
	match_found = TRUE
	terminate = ""
	#
	# If -m option was specified, then match as many patterns on each line
	# as is possible
	#
	if (options ~ /m/)
		{
		one_match = FALSE
		j = 1
		i = 1
		while (match_found && i <= pattern_length)
			{
			if (token[j] !~ pattern[i] && !one_match) match_found = FALSE
			else if (token[j] !~ pattern[i] && one_match)
				{
				j += 1
				one_match = FALSE
				}
			else
				{
				i += 1
				one_match = TRUE
				}
			}
		end_of_pattern = j
		}
	else
		{
		for (i = 1; i <= pattern_length; i += 1)
			{
			if (token[i] !~ pattern[i]) { match_found = FALSE ; break }
			}
		end_of_pattern = i - 1
		}
	#
	# If they match, print a comment showing this; print the correct exclusive
	# interpretations for the first record of the pattern; print the records
	# of the pattern (and any other records in between); and print some
	# spine-path terminators.
	#
	if (match_found)
		{
		data_count = 1
		print "!! Pattern found at line " \
		      FNR - (buffer_length-pattern_start) " of file " file_name
		if (past_interp[interp_tail] != "") print past_interp[interp_tail]
		if (past_tand_interp[tand_interp_tail] != "") print past_tand_interp[tand_interp_tail]	# NEWLY ADDED
		for (x in current_tand_interp) print current_tand_interp[x]	# NEWLY ADDED
		#
		# Print out only the data records that were matched by the pattern
		# (may be less than pattern_length because of -x option)
		#
		for (i = pattern_start; i <= buffer_length; i += 1)
			{
			if (data_count > end_of_pattern) break
			else if (buffer[i] ~ /^\*/) print buffer[i]
			else if (buffer[i] ~ /^!/ && options !~ /c/) print buffer[i]
			else if (options ~ /s/ && buffer[i] ~ skip_reg) print buffer[i]
			else
				{
				print buffer[i]
				data_count += 1
				}
			}
		#
		# Determine how many spines were active during the last data record
		# and print out the appropriate number of spine-path terminators
		#
		i -= 1
		while (buffer[i] ~ /^!!/ && i >= pattern_start) i -= 1
		if (i < pattern_start || buffer[i] ~ spine_terminate) spine_count = 0
		else spine_count = split(buffer[i],temp_array,"\t")
		if (spine_count > 0) terminate = "*-"
		for (i = 2; i <= spine_count; i += 1)
			terminate = terminate "\t*-"
		if (terminate != "") print terminate
		}
	#
	# Determine where the next possible pattern could start and then reset
	# the token array
	#
	i = pattern_start + 1
	while ((buffer[i] ~ /^[*!]/) || (options ~ /s/ && buffer[i] ~ skip_reg))
		{
		if (options ~ /c/ && buffer[i] ~ /^!/) break
		i += 1
		}
	buffer_start = pattern_start = i
	for (i = 1; i <= token_length; i += 1) token[i] = token[i+1]
	token_length -= 1
	}

#########################################################################
##				Function process_no_option
#	This function will process the patterns as specified when neither the
# -t nor the -e options were specified.
#
function process_no_option(   i,j,one_match,match_found)
	{
	#
	# Loop through the arrays to look for a pattern
	#
	match_found = TRUE
	#
	# If -m option was specified, then match as many patterns on each line
	# as is possible
	#
	if (options ~ /m/)
		{
		one_match = FALSE
		j = 1
		i = 1
		while (match_found && i <= pattern_length)
			{
			if (token[j] !~ pattern[i] && !one_match) match_found = FALSE
			else if (token[j] !~ pattern[i] && one_match)
				{
				j += 1
				one_match = FALSE
				}
			else
				{
				i += 1
				one_match = TRUE
				}
			}
		}
	else
		{
		for (i = 1; i <= pattern_length; i += 1)
			{
			if (token[i] !~ pattern[i]) { match_found = FALSE ; break }
			}
		}
	#
	# If a match was found, print a comment showing this
	#
	if (match_found)
		print "!! Pattern found at line " \
		      FNR - (buffer_length-pattern_start) " of file " file_name
	#
	# Determine where the next possible pattern could start and then reset
	# the token array
	#
	i = pattern_start + 1
	while ((buffer[i] ~ /^[*!]/) || (options ~ /s/ && buffer[i] ~ skip_reg))
		{
		if (options ~ /c/ && buffer[i] ~ /^!/) break
		i += 1
		}
	buffer_start = pattern_start = i
	for (i = 1; i <= token_length; i += 1) token[i] = token[i+1]
	token_length -= 1
	}

################################################################
##			Function Store_new_interps
#	This function stores the new interpretations found in an
# interpretation record for each spine.
#
function store_new_interps(      j)
	{
	#
	# Loop through each valid spine
	#	
	for (j = 1; j <= current_no_of_spines; j++)
		{
		#
		# Store exclusive interpretations in current_interp[]
		#
		if ($j ~ /^\*\*/) current_interp[j] = $j
		else if ($j ~ /^\*/) current_tand_interp[j] = $j	# NEWLY ADDED
		}
	}

##############################################################
##			Function Store_indicators
#	This function allows the spine-path indicators for the 
# current record to be stored in the array 'path_indicator' so
# that they may be used later.
#
function store_indicators(   i)
	{
	for (i = 1; i <= current_no_of_spines; i++)
		{
		#
		# If the current field is a null interpretation
		#
		if ($i == "*") path_indicator[i] = "*"
		#
		# If the current field is a spine-path indicator, store it
		#
		else path_indicator[i] = substr($i,2,1)
		}
	} 

##################################################################
##				Function Process_indicators
#	This function takes the spine-path indicators that were stored
# in the array 'path_indicator' in the function 'store_indicators'
# and manipulates the arrays 'path_indicator' and 'current_interp'
# according to the contents of the array 'path_indicator'.
#
function process_indicators(    i)
	{
	i = 1
	#
	# Loop through the entire array of 'path_indicator'
	# NOTE: This array may be altered during the process
	#
	while (i <= current_no_of_spines)
		{
		#
		# If a spine-split indicator occurred
		#
		if (path_indicator[i] == "^")
			{
			#
			# Insert a new array position in 'path_indicator'
			# and 'current_interp'
			#
			ins_array_pos(i+1)
			path_indicator[i+1] = "*"
			current_interp[i+1] = current_interp[i]	# Copy interpretations to the new spine.
			current_tand_interp[i+1] = current_tand_interp[i]	# Copy tandem interpretations to the new spine.	# NEWLY ADDED.
			}
		#
		# If a spine-termination indicator occurred
		#
		if (path_indicator[i] == "-")
			{
			#
			# Delete that spine from the arrays and look at the same
			# position the next time through the loop
			#
			del_array_pos(i)
			i--
			}
		#
		# If a spine-exchange indicator occurred
		#
		if (path_indicator[i] == "x")
			{
			#
			# Exchange the paths
			#
			exchange_spines()
			}
		#
		# If a join-path indicator occurred
		#
		if (path_indicator[i] == "v")
			{
			del_array_pos(i+1)	# Delete NEXT spine (must be join-path
							# spine according to humdrum).
			#
			# If the next indicator is not a join-path indicator
			# then return to original conditions
			#
			if (path_indicator[i+1] != "v")
				{
				path_indicator[i] = "*"
				}
			#
			# Otherwise, look at current and next paths to join
			#
			else i--
			}
		#
		# If a new-path indicator occurred
		#
		if (path_indicator[i] == "+")
			{
			#
			# Insert a new position in the arrays
			#
			ins_array_pos(i+1)
			current_interp[i+1] = ""
			current_tand_interp[i+1] = ""	# NEWLY ADDED.
			}
		i++
		}
	}

############################################################
##			Function Ins_array_pos
#	This function inserts a new position in the arrays 'path_indicator'
# and 'current_interp' and copies elements so that order is preserved
#
function ins_array_pos(current_element,     j) 
	{
	for (j = current_no_of_spines+1; j > current_element; j--)
		{
		path_indicator[j] = path_indicator[j-1]
		current_interp[j] = current_interp[j-1]	
		current_tand_interp[j] = current_tand_interp[j-1]	# NEWLY ADDED.
		}
	path_indicator[current_element-1] = "*"	# (Avoid possibility of
	path_indicator[current_element] = "*"	#  reprocessing the old spine.)
	current_no_of_spines++
	}

##############################################################
##			Function Del_array_pos
#	Performs the opposite of function 'ins_array_pos'.
#
function del_array_pos(current_element,     j)
	{
	for (j = current_element; j < current_no_of_spines; j++)
		{
		path_indicator[j] = path_indicator[j+1]
		current_interp[j] = current_interp[j+1]
		current_tand_interp[j] = current_tand_interp[j+1]	# NEWLY ADDED.
		}
	delete path_indicator[j]
	delete current_interp[j]
	delete current_tand_interp[j]
	current_no_of_spines--
	} 

###################################################################
##				Function Exchange_spines
#	This function exhanges two spines by exchanging the corresponding
# elements in 'current_interp'
#
function exchange_spines(    j,count,arraya,temp)
	{
	count = 1
	# Determine which spines are exchanged
	for (j = 1; j <= current_no_of_spines; j += 1)
		{
		if (path_indicator[j] == "x")
			{
			arraya[count] = j
			count += 1
			path_indicator[j] = "*"
			}
		}
	# Exchange interpretations between the two spines
	temp = current_interp[arraya[1]]
	current_interp[arraya[1]] = current_interp[arraya[2]]
	current_interp[arraya[2]] = temp

	# Similarly, exchange tandem interpretations.	# NEWLY ADDED.
	temp = current_tand_interp[arraya[1]]	# NEWLY ADDED.
	current_tand_interp[arraya[1]] = current_interp[arraya[2]]	# NEWLY ADDED.
	current_tand_interp[arraya[2]] = temp	# NEWLY ADDED.
	}
